[SPARK-20642][core] Store FsHistoryProvider listing data in a KVStore.

vanzin opened the pull request:
The application listing is still generated from event logs, but is now stored
in a KVStore instance. By default an in-memory store is used, but a new config
allows setting a local disk path to store the data, in which case a LevelDB
store will be created.

The provider stores things internally using the public REST API types; I believe
this is better going forward since it will make it easier to get rid of the
internal history server API which is mostly redundant at this point.

I also added a finalizer to LevelDBIterator, to make sure that resources are
eventually released. This helps when code iterates but does not exhaust the
iterator, thus not triggering the auto-close code.

HistoryServerSuite was modified to not re-start the history server unnecessarily;
this makes the json validation tests run more quickly.

Commit:
SHS-NG M2: Store FsHistoryProvider listing data in a KVStore.

The application listing is still generated from event logs, but is now stored
in a KVStore instance. By default an in-memory store is used, but a new config
allows setting a local disk path to store the data, in which case a LevelDB
store will be created.

The provider stores things internally using the public REST API types; I believe
this is better going forward since it will make it easier to get rid of the
internal history server API which is mostly redundant at this point.

I also added a finalizer to LevelDBIterator, to make sure that resources are
eventually released. This helps when code iterates but does not exhaust the
iterator, thus not triggering the auto-close code.

HistoryServerSuite was modified to not re-start the history server unnecessarily;
this makes the json validation tests run more quickly.

vanzin commented:
For context:

* Project link: https://issues.apache.org/jira/browse/SPARK-18085
* Upcoming PRs that build on this code: https://github.com/vanzin/spark/pulls

vanzin commented on scalastyle-config.xml:
```diff
@@ -86,7 +86,7 @@ This file is divided into 3 sections:
 
   <check level="error" class="org.scalastyle.scalariform.ObjectNamesChecker" enabled="true">
-    <parameters><parameter name="regex"><![CDATA[[A-Z][A-Za-z]*]]></parameter></parameters>
+    <parameters><parameter name="regex"><![CDATA[(config|[A-Z][A-Za-z]*)]]></parameter></parameters>
```
I'm adding this so that we can have an object named `config`; this seems cleaner than the current approach of having `package object config`, which requires every constant to be tagged as private, instead of just marking the whole object as `private[spark]`.

squito replied:
Agree it makes more sense to make the whole thing `private[spark]`, but couldn't you just turn the rule off around this object, like we do for println?

(Also I'm confused why `package object config` doesn't violate the rule).

vanzin replied:
That could be done, but I want this to be the "official" pattern going forward, and having to disable scalastyle every time it's used is kinda ugly and somewhat detracts from the official-ness of it.

squito replied:
ok makes sense.  I wasn't paying enough attention to the actual change at first, thought you were just turning the rule entirely off, but I agree this changes makes sense.

btw I was curious why `package object config` was OK -- its just a hard-coded special case: https://github.com/scalastyle/scalastyle/blob/master/src/main/scala/org/scalastyle/scalariform/ClassNamesChecker.scala#L74

SparkQA commented:
**[Test build #80420 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80420/testReport)** for PR 18887 at commit [`842589d`](https://github.com/apache/spark/commit/842589d34a2c2b97180d392d35eaa985edbfd671).

Commit:
SHS-NG M2: Re-work LevelDBIterator.finalize().

Make sure the db is still open before trying to close the iterator,
otherwise it may cause a JVM crash.

SparkQA commented:
**[Test build #80422 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80422/testReport)** for PR 18887 at commit [`1f08bd7`](https://github.com/apache/spark/commit/1f08bd7d7f5a57d27a4748baff0353f4a9f285da).

SparkQA commented:
**[Test build #80420 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80420/testReport)** for PR 18887 at commit [`842589d`](https://github.com/apache/spark/commit/842589d34a2c2b97180d392d35eaa985edbfd671).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class KVStoreMetadata(`
  * `case class LogInfo(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80420/
Test PASSed.

Commit:
Add private where needed.

SparkQA commented:
**[Test build #80431 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80431/testReport)** for PR 18887 at commit [`1ec1a67`](https://github.com/apache/spark/commit/1ec1a672673eafa9f04ad81c1fff3feadfcddd2c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80422/
Test PASSed.

SparkQA commented:
**[Test build #80431 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80431/testReport)** for PR 18887 at commit [`1ec1a67`](https://github.com/apache/spark/commit/1ec1a672673eafa9f04ad81c1fff3feadfcddd2c).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80431/
Test FAILed.

vanzin commented:
hmm... @squito @ajbozarth @jerryshao

ajbozarth commented on core/src/main/scala/org/apache/spark/status/api/v1/api.scala:
```diff
@@ -31,6 +33,9 @@ class ApplicationInfo private[spark](
     val attempts: Seq[ApplicationAttemptInfo])
 
+@JsonIgnoreProperties(
+  value = Array("startTimeEpoch", "endTimeEpoch", "lastUpdatedEpoch"),
```
Will this exclude the Epoch values from the api? Because if I remember correctly we added those for the api specifically

vanzin replied:
No, this just avoids trying to deserialize them, which would cause an error because these properties have no setter.

ajbozarth reviewed (commented):
A few questions and feedback, but the SHS stuff looks good, I'm not well versed in the KVStore/DB side of things so I can't vouch for those, though the code looks fine.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala:
```diff
@@ -76,6 +76,14 @@ private[history] case class LoadedAppUI(
 
   /**
+   * The number of applications available for listing. Separate method in case it's cheaper
+   * to get a count than to calculate the whole listing.
```
This is an interface, so this was added to allow implementations to override this method if that makes sense.

It just looks like I lost the override in one of my rebases, so let me add that back.

vanzin replied:
Actually it doesn't seem like this is used anymore and I can remove it...

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -742,53 +698,145 @@ private[history] object FsHistoryProvider {
+  val fileSize: Long)
+
+private[history] class AttemptInfoWrapper(
+    val info: v1.ApplicationAttemptInfo,
```
Yes, I'm using this syntax because in many places there are conflicting type names in the API package and in other packages.

Commit:
Feedback.

SparkQA commented:
**[Test build #80639 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80639/testReport)** for PR 18887 at commit [`b696f96`](https://github.com/apache/spark/commit/b696f963ee001687ecd9266b9115ba2aa2867740).

SparkQA commented:
**[Test build #80639 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80639/testReport)** for PR 18887 at commit [`b696f96`](https://github.com/apache/spark/commit/b696f963ee001687ecd9266b9115ba2aa2867740).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80639/
Test PASSed.

jerryshao commented on docs/monitoring.md:
```diff
@@ -220,6 +220,13 @@ The history server can be configured as follows:
     </td>
   </tr>
+  <tr>
+    <td>spark.history.store.path</td>
```
Check from the log, looks like we don't have a default path for local history store, do we need to add a default value here in doc?

```  
val LOCAL_STORE_DIR = ConfigBuilder("spark.history.store.path")
    .stringConf
    .createOptional
```

vanzin replied:
I'm using this as a trigger of whether to use the disk store or not; if you set a local store directory, you're using the disk store, otherwise you're using the memory store. I do need to update it in the documentation, though.

jerryshao reviewed (commented):
LGTM overall, just some minor comments.

One question about `FsHistoryProvider`, looks like we removed some synchronizations here, I'm not sure if our underlying kvstore well supports concurrency, or it is not an issue any more?

Commit:
Feedback.

SparkQA commented:
**[Test build #80701 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80701/testReport)** for PR 18887 at commit [`519dab0`](https://github.com/apache/spark/commit/519dab056964dae71309f65bcadee8ec08366284).

SparkQA commented:
**[Test build #80701 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80701/testReport)** for PR 18887 at commit [`519dab0`](https://github.com/apache/spark/commit/519dab056964dae71309f65bcadee8ec08366284).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80701/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -301,6 +334,7 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
   }
 
   override def stop(): Unit = {
+    listing.close()
```
if this throws an exception, should we still try to cleanup `initThread`?

squito commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -117,17 +122,37 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
 
-  // List of application logs to be deleted by event log cleaner.
-  private var attemptsToClean = new mutable.ListBuffer[FsApplicationAttemptInfo]
+  private val listing = storePath.map { path =>
```
given the large initializer, could you add an explicit type annotation to `listing` to help the reader?

squito commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -316,25 +350,22 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+            SparkHadoopUtil.get.checkAccessPermission(entry, FsAction.READ) &&
+            recordedFileSize(entry.getPath()) < entry.getLen()
         }
         .flatMap { entry => Some(entry) }
```
realize this isn't your change, but what is the point of this?  isn't it a no-op?

squito commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -742,53 +696,146 @@ private[history] object FsHistoryProvider {
+    var coresPerExecutor: Option[Int] = None
+    var memoryPerExecutorMB: Option[Int] = None
+
+    def toView(attempts: List[AttemptInfoWrapper]): ApplicationInfoWrapper = {
```
looks like this is only ever called with one `AttemptInfoWrapper`, so simpler if you remove the `List`.

squito reviewed (commented):
I have one big question -- do you think that in the initial version this should replace the old FSHistoryProvider?  I feel like we should have one version where the old code is still available, controlled by a feature flag.

Other than that mostly minor things on the code.

Commit:
Feedback.

SparkQA commented:
**[Test build #80750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80750/testReport)** for PR 18887 at commit [`dc642bd`](https://github.com/apache/spark/commit/dc642bd70042da965387916656747ae78acdc192).

vanzin commented:
> I feel like we should have one version where the old code is still available, controlled by a feature flag.

I'm not sure exactly what you're suggesting. The default behavior is still, as much as possible, the old one: everything is kept in memory. Keeping the exact old code in place would mean forking `FsHistoryProvider`, which is not something I see as desirable.

Also, forgot to reply to an earlier comment by @jerryshao : 

> looks like we removed some synchronizations here, I'm not sure if 

KVStore instances are thread safe, so a lot of the old synchronization in this class does not apply. There was code before that used non-thread-safe data structures (such as `attemptsToClean`), which has been replaced with storing things in the KVStore instance instead.

squito commented:
>> I feel like we should have one version where the old code is still available, controlled by a feature flag.

>I'm not sure exactly what you're suggesting. The default behavior is still, as much as possible, the old one: everything is kept in memory. Keeping the exact old code in place would mean forking FsHistoryProvider, which is not something I see as desirable.

long-term, definitely not desirable.  My concern is that there is a lot of new code taking effect here, on some important functionality -- we don't want some bug to prevent adoption of 2.3. For one version, you could leave the old one available, rename it to something else, and put a big disclaimer in the code that its obsolete, and as long as thing are smooth for 2.3 delete it entirely for 2.4.  The HS isn't as core or tricky as the network module but I'm thinking of this like netty vs. nio.

vanzin commented:
There's quite a log of unit tests that cover this code; forking would mean also making those unit tests run against both versions of the code so that no one breaks anything, and potentially fixing bugs in two different places if they're found.

I'd rather avoid the overhead. The bulk of `FsHistoryProvider`, which is the part that actually monitors the file system and makes decisions about when to parse things, is mostly unchanged.

And the rest of this project is way more disruptive than this one change.

SparkQA commented:
**[Test build #80750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80750/testReport)** for PR 18887 at commit [`dc642bd`](https://github.com/apache/spark/commit/dc642bd70042da965387916656747ae78acdc192).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80750/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #80757 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80757/testReport)** for PR 18887 at commit [`dc642bd`](https://github.com/apache/spark/commit/dc642bd70042da965387916656747ae78acdc192).

SparkQA commented:
**[Test build #80757 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80757/testReport)** for PR 18887 at commit [`dc642bd`](https://github.com/apache/spark/commit/dc642bd70042da965387916656747ae78acdc192).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80757/
Test PASSed.

squito commented:
Just to keep others in the loop, Marcelo and I talked about this some offline.  I think this PR itself is fine, but to me this is an important point in the larger history server project he's doing, so I'm going to take a look at the rest of the changes as well before I feel comfortable merging.  Also he explained why it extremely complicated to keep both the old & new version available, which make sense to me, though I may poke at a version myself just to see how complex it is.

squito commented:
I futzed around for a while with trying to keep the old stuff around, and I realized it really would be quite a mess.  The biggest problem is that the old rest api is just waay too tied into the UI, eg https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/api/v1/ExecutorListResource.scala#L26 , probably my fault in that implementation of the rest apis.  You could keep the old code around, but would involve so much moving and refactoring that it seems pointless.

I'm still looking at the other changes in the larger project, and I'd encourage other reviewers to do the same.

gatorsmile commented:
I have the same concern, as @squito said above https://github.com/apache/spark/pull/18887#pullrequestreview-56700674. Refactoring the old codes does not seem pointless to me. 

To evaluate the impacts, I have a few questions:
- What is the migration proposal/guides? 
- What is the proposal when users have multiple different Spark versions in their production environment or migration procedure? 
- How users should do when users hit a serious bug we are unable to find at this stage? 
- What are the external behavior changes between the new and the old ones?

vanzin commented:
> What is the migration proposal/guides?

Not sure what you mean. There's no change in behavior by default, so there's no migration of anything needed.

> What should users do when they hit serious bugs that we are unable to find at this stage?

This is always a question when you introduce a change. This particular project makes it a bit worse since it's a bigger change. But the answer is the same: file a bug, we fix it, next release has the fix; if it's really important for you, you can patch your own Spark, or revert to an older release, as many people do when they find blockers.

The changes in this project do not influence whether your application will fail or not; they're isolated to the UI, so at worst it will make things harder to debug until the issues are fixed.

On the other hand, a lot can be mitigated by testing; there's already a lot of test coverage for parts of this code, but the UI is kinda the exception there. So the longer these changes stay in review instead of being committed, the less coverage they'll have in people's day to day testing, actually increasing the risk that some bug might be missed.

> What are the external behavior changes between the new and the old ones?

None by default.

gatorsmile reviewed (changes requested):
I quickly went over the changes and left a few comments. I might need more time to understand the potential impact.

gatorsmile commented on core/src/main/scala/org/apache/spark/deploy/history/config.scala:
```diff
@@ -0,0 +1,46 @@
+    .timeConf(TimeUnit.SECONDS)
+    .createWithDefaultString("7d")
+
+  val LOCAL_STORE_DIR = ConfigBuilder("spark.history.store.path")
```
It'd better to document the default one is an in-memory store.

vanzin replied:
Correct, other parameters are the same.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -117,17 +122,37 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+      } else if (meta.version != CURRENT_LISTING_VERSION || !logDir.equals(meta.logDir)) {
+        logInfo("Detected mismatched config in existing DB, deleting...")
+        db.close()
+        Utils.deleteRecursively(dbPath)
```
Yes; the code will re-create the data from the event logs when that happens.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -301,9 +334,13 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+        initThread.join()
+      }
+    } finally {
+      listing.close()
```
It might leave JNI handles open (a.k.a. a memory leak).

Also, this doesn't apply to this change, but later when the UI info is also written to disk, it could prevent the UI db from being replaced with an updated one, since its files would be opened.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -422,208 +455,101 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
-      // Use loading time as lastUpdated since some filesystems don't update modifiedTime
-      // each time file is updated. However use modifiedTime for completed jobs so lastUpdated
-      // won't change whenever HistoryServer restarts and reloads the file.
-      val lastUpdated = if (appCompleted) fileStatus.getModificationTime else clock.getTimeMillis()
```
Yes, the behavior is the same. It's now handled in `AppListingListener` (`onApplicationStart` and `onApplicationEnd` callbacks).

I don't see an explicit test for this behavior.

vanzin replied:
Actually I think ""log cleaner for inProgress files" might test this.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -742,53 +698,146 @@ private[history] object FsHistoryProvider {
   private val LOG_START_EVENT_PREFIX = "{\"Event\":\"SparkListenerLogStart\""
+
+  /** Current version of the data written to the listing database. */
+  private val CURRENT_LISTING_VERSION = 1L
```
I'll add a more verbose comment; but this is basically me punting proper versioning to the next Spark release after this code is added. This version number is a "nuclear option"; if we break the data that is serialized to disk, we increase this number, and the new SHS will delete all old data and re-generated it from event logs.

I'm punting because there is no versioning issue in the first version; there's no existing data that the SHS might try to read.

I plan to take a closer look at versioning after all the initial PRs go in, but leaving this here gives us a choice in case there's a more urgent need to break things.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -742,53 +698,146 @@ private[history] object FsHistoryProvider {
+
+    attempt.attemptId = event.appAttemptId
+    attempt.startTime = new Date(event.time)
+    attempt.lastUpdated = new Date(clock.getTimeMillis())
```
Both are basically `System.currentTimeMillis()`, so yes.

Commit:
More feedback.

SparkQA commented:
**[Test build #81697 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81697/testReport)** for PR 18887 at commit [`9020184`](https://github.com/apache/spark/commit/9020184bba90fc1c7394ae8ab91877efe0699914).

SparkQA commented:
**[Test build #81697 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81697/testReport)** for PR 18887 at commit [`9020184`](https://github.com/apache/spark/commit/9020184bba90fc1c7394ae8ab91877efe0699914).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81697/
Test PASSed.

gatorsmile commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -720,19 +633,67 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+   * where two threads are processing separate attempts of the same application.
+   */
+  private def addListing(app: ApplicationInfoWrapper): Unit = listing.synchronized {
+    val attempt = app.attempts.head
```
Why `.head`?

vanzin replied:
Because the listener processes a single event log; one event log == one attempt.

cloud-fan replied:
shall we add an assert to make sure we only have one attempt here?

gatorsmile reviewed (commented):
Just finish another pass. Thanks!

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -624,7 +639,9 @@ class FsHistoryProviderSuite extends SparkFunSuite with BeforeAndAfter with Matc
-    new SparkConf().set("spark.history.fs.logDirectory", testDir.getAbsolutePath())
+    new SparkConf()
+      .set("spark.history.fs.logDirectory", testDir.getAbsolutePath())
+      .set(LOCAL_STORE_DIR, Utils.createTempDir().getAbsolutePath())
```
The `kvstore` module already tests both in-memory and disk stores extensively (running the same set of tests on both). There isn't really anything that testing both here can add to those tests.

At most I could make `FsHistoryProvider` do round-robin between disk and in-memory; but I don't see the point of running these tests twice.

gatorsmile replied:
The test cases in `kvstore` are just unit test cases. We also need integration tests for ensuring they work as expected.

vanzin replied:
Disagree. The code in the SHS, other than instantiation, is exactly the same regardless of the underlying store implementation. Meaning that if both stores are tested equally, this code should also work equally there.

By testing the disk store only, we're testing all the code involved here (which is the interaction with the store itself + serialization, which is not tested by the in-memory store).

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/HistoryServerSuite.scala:
```diff
@@ -74,6 +76,7 @@ class HistoryServerSuite extends SparkFunSuite with BeforeAndAfter with Matchers
       .set("spark.history.fs.logDirectory", logDir)
       .set("spark.history.fs.update.interval", "0")
       .set("spark.testing", "true")
+      .set(LOCAL_STORE_DIR, storeDir.getAbsolutePath())
```
See previous reply.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -720,19 +633,67 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    val newAppInfo = new ApplicationInfoWrapper(
+      app.info,
+      attempts.sortWith(compareAttemptInfo))
+    listing.write(newAppInfo)
```
Yes.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -720,19 +633,67 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
 private[history] object FsHistoryProvider {
-  val DEFAULT_LOG_DIR = "file:/tmp/spark-events"
-
   private val NOT_STARTED = "<Not Started>"
```
This existed for compatibility with really old event logs that 2.x does not support anymore.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -742,53 +703,150 @@ private[history] object FsHistoryProvider {
+  private val attempt = new MutableAttemptInfo(log.getPath().getName(), log.getLen())
+
+  override def onApplicationStart(event: SparkListenerApplicationStart): Unit = {
+    app.id = event.appId.orNull
```
No; Spark 2.x does not support applications without an id, so an event log that doesn't have an app id is an invalid one.

vanzin replied:
Yes. And 1.6 always writes an app id.

Only really, really old logs don't have app ids, and those are not supported. You can scavenge git / jira history if you're really interested in when that happens.

Commit:
Cleanup.

Commit:
Merge branch 'master' into SPARK-20642

Commit:
Add test for CURRENT_LISTING_VERSION.

SparkQA commented:
**[Test build #81890 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81890/testReport)** for PR 18887 at commit [`56b68a0`](https://github.com/apache/spark/commit/56b68a06e5fc41e423a10ed767c1eb50d62395f5).

SparkQA commented:
**[Test build #81890 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81890/testReport)** for PR 18887 at commit [`56b68a0`](https://github.com/apache/spark/commit/56b68a06e5fc41e423a10ed767c1eb50d62395f5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81890/
Test PASSed.

cloud-fan commented:
can you also include the format? i.e. how to convert the list information to key-value entries

vanzin commented:
@cloud-fan not sure I understand your question. What format are you talking about?

cloud-fan commented:
nvm, I looked at your previous PRs, the "format" is already defined in kv-store, as the kv-store can read/write objects directly.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -422,208 +457,101 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+        .closeableIterator())
+
+      iterator.get.asScala.foreach { app =>
+        val (remaining, toDelete) = app.attempts.partition { attempt =>
```
You're just repeating the method's javadoc.

vanzin replied:
Ah, you're talking about attempts, not applications.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -422,208 +457,101 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+          val newApp = new ApplicationInfoWrapper(app.info, remaining)
+          listing.write(newApp)
+        } else {
+          listing.delete(app.getClass(), app.id)
```
Sure. But that doesn't really make it safe (e.g. what if deleting one attempt succeeds, but deleting a second one fails? You'd end up with pointers to the failed attempt, unless you update the listing after each deletion; and if updating the listing failed, you still end up in the same situation.)

Later milestones have code that recovers from those situations, though.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -720,75 +634,218 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    val info: v1.ApplicationInfo,
+    val attempts: List[AttemptInfoWrapper]) {
+
+  @JsonIgnore @KVIndexParam
```
All objects written to a kvstore need a "natural" (= unnamed) index.

Commit:
Feedback.

Commit:
Merge branch 'master' into SPARK-20642

SparkQA commented:
**[Test build #82154 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82154/testReport)** for PR 18887 at commit [`b2b969c`](https://github.com/apache/spark/commit/b2b969cf49579b26b37b23757a55f1e7808330e5).

Commit:
Run single test with in-memory store.

SparkQA commented:
**[Test build #82155 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82155/testReport)** for PR 18887 at commit [`cd2172a`](https://github.com/apache/spark/commit/cd2172a027cba92755c89dbaa5f010917749a98a).

SparkQA commented:
**[Test build #82154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82154/testReport)** for PR 18887 at commit [`b2b969c`](https://github.com/apache/spark/commit/b2b969cf49579b26b37b23757a55f1e7808330e5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82154/
Test PASSed.

SparkQA commented:
**[Test build #82155 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82155/testReport)** for PR 18887 at commit [`cd2172a`](https://github.com/apache/spark/commit/cd2172a027cba92755c89dbaa5f010917749a98a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82155/
Test PASSed.

gatorsmile commented:
LGTM

gatorsmile commented:
Thanks for your work!

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -722,75 +640,215 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+  @KVIndexParam logPath: String,
+  fileSize: Long)
+
+private[history] class AttemptInfoWrapper(
```
why this one doesn't have an index?

vanzin replied:
Because it's not directly written to the store.

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/config.scala:
```diff
@@ -0,0 +1,48 @@
+    .stringConf
+    .createWithDefault(DEFAULT_LOG_DIR)
+
+  val MAX_LOG_AGE_S = ConfigBuilder("spark.history.fs.cleaner.maxAge")
```
what does the ending `S` mean? `seconds`?

vanzin replied:
Yes. This pattern is used in a bunch of other places to indicate the unit of time of the config.

cloud-fan commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -496,7 +517,7 @@ class FsHistoryProviderSuite extends SparkFunSuite with BeforeAndAfter with Matc
       var provider: FsHistoryProvider = null
       try {
-        provider = new FsHistoryProvider(conf)
+        provider = newProvider(conf)
```
It seems worse to have this `newProvider` method and create a lot of diff.

cloud-fan commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -496,7 +517,7 @@ class FsHistoryProviderSuite extends SparkFunSuite with BeforeAndAfter with Matc
       var provider: FsHistoryProvider = null
       try {
-        provider = new FsHistoryProvider(conf)
+        provider = newProvider(conf)
```
it seems worse to have `newProvider` method and create a lof of diff.

vanzin replied:
This was part of some code I only partially reverted. Will revert these.

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -243,42 +282,38 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
         }
-      }
     } catch {
-      case e: FileNotFoundException => None
```
we will never hit `FileNotFoundException`?

vanzin replied:
I guess it still can until I remove the current replay code in a later change.

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -243,42 +282,38 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
-        appInfo.attempts.find(_.attemptId == attemptId).flatMap { attempt =>
+      val appInfo = load(appId)
+      appInfo.attempts
+        .find { attempt => attempt.info.attemptId == attemptId }
```
nit: `_.info.attemptId == attemptId`

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -243,42 +282,38 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
-              Some(attempt.lastUpdated), attempt.startTime)
+            SparkUI.createHistoryUI(conf, replayBus, appSecManager, appInfo.info.name,
+              HistoryServer.getAttemptURI(appId, attempt.info.attemptId),
+              Some(attempt.info.lastUpdated.getTime()), attempt.info.startTime.getTime())
```
Shall we make `load` return `ApplicationHistoryInfo` instead of `ApplicationInfoWrapper`? Then we can reduce a lot of unnecessary diff.

vanzin replied:
`AttemptInfoWrapper` stores information that is not available in the public API and is used by the provider. Later changes also add more fields to `ApplicationInfoWrapper` that are needed by the provider.

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -243,42 +282,38 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
-            None
-          }
-
+          ui.appSparkVersion = appListener.appSparkVersion.getOrElse("")
```
add an `assert(appListener.appId.isDefined)`?

vanzin replied:
If the listing exists it's unlikely that this would ever trigger, but sure.

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -318,25 +357,21 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+      val statusList = Option(fs.listStatus(new Path(logDir))).map(_.toSeq).getOrElse(Nil)
       // scan for modified applications, replay and merge them
-      val logInfos: Seq[FileStatus] = statusList
+      val logInfos = statusList
```
can we just do `val logInfos = Option(fs.listStatus(new Path(logDir))).map(_.toSeq).getOrElse(Nil)`?

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -722,75 +640,215 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+  private def addListing(app: ApplicationInfoWrapper): Unit = listing.synchronized {
+    val attempt = app.attempts.head
+
+    val oldApp = try {
```
you can call `load` here

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -722,75 +640,215 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+        app
+    }
+
+    def compareAttemptInfo(a1: AttemptInfoWrapper, a2: AttemptInfoWrapper): Boolean = {
```
can we add a `isStartedEarlierThan(other)` method in `AttemptInfoWrapper`?

vanzin replied:
This is only used here, I'd rather keep the logic local.

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -722,75 +640,215 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    val attempt = app.attempts.head
+
+    val oldApp = try {
+      listing.read(classOf[ApplicationInfoWrapper], app.id)
```
you can call `load` here

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -424,208 +459,105 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
 
   /**
-   * Replay the log files in the list and merge the list of old applications with new ones
+   * Replay the given log file, saving the application in the listing db.
```
why update the comment? We don't merge any more?

vanzin replied:
No. If you look at the old code it did a "merge sort" kinda thing to create an updated listing. `KVStore` sorts things internally so there's no need for that code anymore - you just write something to it, and it's sorted magically.

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -424,208 +459,105 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+      iterator.get.asScala.foreach { app =>
+        // Applications may have multiple attempts, some of which may not need to be deleted yet.
+        val (remaining, toDelete) = app.attempts.partition { attempt =>
+          attempt.info.lastUpdated.getTime() >= maxTime
```
nit: previously we use `>` not `>=`, see https://github.com/apache/spark/pull/18887/files#diff-a7befb99e7bd7e3ab5c46c2568aa5b3eL561

cloud-fan commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -685,26 +618,11 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
-      appInfo.attempts.find(_.attemptId == attemptId)
-    }
+    val count = listing.count(classOf[ApplicationInfoWrapper])
+    s"""|FsHistoryProvider{logdir=$logDir,
```
one space after `|`

vanzin replied:
That would make the string look weird.

Commit:
Get rid of `newProvider`.

SparkQA commented:
**[Test build #82206 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82206/testReport)** for PR 18887 at commit [`5eff2c5`](https://github.com/apache/spark/commit/5eff2c5f0891c6ea3fcfbbc7dacbbaf56c1d1788).

SparkQA commented:
**[Test build #82206 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82206/testReport)** for PR 18887 at commit [`5eff2c5`](https://github.com/apache/spark/commit/5eff2c5f0891c6ea3fcfbbc7dacbbaf56c1d1788).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82206/
Test PASSed.

cloud-fan commented:
LGTM, merging to master!

mgaido91 commented:
sorry, this is causing a checkstyle error for me while building the project due to the presence of a finalizer method. I can't understand how this was able to pass the Jenkins test. Does someone have any clue on that? Thanks.

cloud-fan commented:
hmmm weird, @srowen any ideas?

srowen commented:
Oh, is it because SBT doesn't run checkstyle but Maven does, and SBT runs the PR builder? This could be my fault, for recently re-enabling checkstyle. This may be why. Well, for now, I'd at least just submit the hotfix to suppress this checkstyle rule

cloud-fan commented:
yea the PR build use SBT. But I think SBT also run checkstyle before?

srowen commented:
I don't think so because checkstyle is for Java only and SBT won't compile Java code.
I'd kind of love to stop mixing SBT and Maven but that's quite a separate thing.

I'll put up a HOTFIX that both fixes the style violation and disables checkstyle again. Now I remember this was the reason -- it wasn't because we used to compile Java code separately from the Scala code.
